### PR TITLE
Parser support for HASH partitioning

### PR DESCRIPTION
**Pull request by adalbertgerald**
Parser support has been added for the hash partitioning clause.

Syntax:
CREATE TABLE <tablename> (attr1 type1, ... , attrn typen)
PARTITION BY HASH (attr(s))
PARTITIONS n;

### CODE REVIEW

**Comment by zuyu** [comment]
@adalbertgerald Please refresh files in `/parser/preprocessed` as well.

**Comment by adalbertgerald** [comment]
Hi @rogersjeffreyl, Can you please review this PR?

File: parser/ParsePartitionClause.hpp [hunk 1]
```diff
@@ -0,0 +1,132 @@
+/**
+ *   Copyright 2016, Quickstep Research Group, Computer Sciences Department,
+ *   University of Wisconsin—Madison.
```
Comment (zuyu) [inline parser/ParsePartitionClause.hpp]: Add two more whitespace indents.

File: parser/ParsePartitionClause.hpp [hunk 2]
```diff
@@ -0,0 +1,132 @@
...
+      if (attribute_name_list_.get() != nullptr) {
+        container_child_field_names->push_back("attribute_name_list");
+        container_child_fields->emplace_back();
+        for (const ParseString& attribute_name : *attribute_name_list_) {
+          container_child_fields->back().push_back(&attribute_name);
+        }
+      }
+
+      non_container_child_field_names->push_back("Number of Partitions");
+      non_container_child_fields->push_back(num_partitions_.get());
```
Comment (zuyu) [inline parser/ParsePartitionClause.hpp]: Reduce two whitespace indents in this method.

File: parser/ParsePartitionClause.hpp [hunk 3]
```diff
@@ -0,0 +1,132 @@
...
+  void getFieldStringItems(std::vector<std::string> *inline_field_names,
+                           std::vector<std::string> *inline_field_values,
+                           std::vector<std::string> *non_container_child_field_names,
+                           std::vector<const ParseTreeNode*> *non_container_child_fields,
+                           std::vector<std::string> *container_child_field_names,
+                           std::vector<std::vector<const ParseTreeNode*>> *container_child_fields) const override {
+      inline_field_names->push_back("partition_type");
+      inline_field_values->push_back(partition_type_->value());
+
+      if (attribute_name_list_.get() != nullptr) {
```
Comment (zuyu) [inline parser/ParsePartitionClause.hpp]: Remove `.get()`.

File: parser/ParsePartitionClause.hpp [hunk 4]
```diff
@@ -0,0 +1,132 @@
...
+        }
+      }
+
+      non_container_child_field_names->push_back("Number of Partitions");
+      non_container_child_fields->push_back(num_partitions_.get());
+  }
+
+ private:
+  std::unique_ptr<ParseString> partition_type_;
+  std::unique_ptr<PtrList<ParseString> > attribute_name_list_;
```
Comment (zuyu) [inline parser/ParsePartitionClause.hpp]: Remove the whitespace between two `>`s.

File: parser/ParsePartitionClause.hpp (hunk 1 above)
Comment (adalbertgerald) [inline parser/ParsePartitionClause.hpp]: @zuyu, can you please let me know why should we add 2 more indents here?

File: parser/ParsePartitionClause.hpp (hunk 2 above)
Comment (adalbertgerald) [inline parser/ParsePartitionClause.hpp]: done.

File: parser/ParsePartitionClause.hpp (hunk 3 above)
Comment (adalbertgerald) [inline parser/ParsePartitionClause.hpp]: I prefer get() with unique pointers. Can you please let me know why should this be removed?

File: parser/ParsePartitionClause.hpp (hunk 4 above)
Comment (adalbertgerald) [inline parser/ParsePartitionClause.hpp]: done.

**Comment by zuyu** [comment]
@adalbertgerald You need to add test cases for parsing partition schemes.

File: parser/ParsePartitionClause.hpp (hunk 1 above)
Comment (zuyu) [inline parser/ParsePartitionClause.hpp]: Because this is not a new line, but a part of previous line.

File: parser/ParsePartitionClause.hpp (hunk 3 above)
Comment (zuyu) [inline parser/ParsePartitionClause.hpp]: Because `unique_ptr` supports `operator !=`.

File: parser/SqlParser.ypp [hunk 5]
```diff
@@ -610,8 +621,8 @@ alter_table_statement:
   };
 
 create_table_statement:
-  TOKEN_CREATE TOKEN_TABLE any_name '(' column_def_commalist ')' opt_table_constraint_def_commalist opt_block_properties {
-    $$ = new quickstep::ParseStatementCreateTable(@1.first_line, @1.first_column, $3, $5, $8);
+  TOKEN_CREATE TOKEN_TABLE any_name '(' column_def_commalist ')' opt_table_constraint_def_commalist opt_block_properties opt_partition_clause {
+    $$ = new quickstep::ParseStatementCreateTable(@1.first_line, @1.first_column, $3, $5, $8, $9);
```
Comment (rogersjeffreyl) [inline parser/SqlParser.ypp]: Shouldn't opt_partition_clause be a part of the create table statement?

**Comment by pateljm** [comment]
Thanks @adalbertgerald for this feature! Thanks @zuyu for your excellent comments. @adalbertgerald Let's have @zuyu work with you on closing this PR, so changing the assignee to @zuyu. 

Cheers,
Jignesh

**Comment by zuyu** [comment]
@adalbertgerald Please sync with the master. Thanks!

File: parser/tests/Create.test [hunk 6]
```diff
@@ -139,6 +139,62 @@ CREATE TABLE test (attr INT) CHECK(attr>0)
                              ^
 ==
 
+# Empty partitioning attribute is not allowed
```
Comment (zuyu) [inline parser/tests/Create.test]: Could you please add the correct queries with hash- and range- partitions first?

Comment (adalbertgerald) [inline parser/tests/Create.test]: @zuyu, do you know how to print the query plan as below for correct statements?

==                                                                                                                                                                                
                                                                                                                                                                                  
CREATE TABLE test (attr INT NULL, attr FLOAT NOT NULL)                                                                                                                            
--                                                                                                                                                                                
CreateTableStatement[relation_name=test]                                                                                                                                          
+-attribute_list=                                                                                                                                                                 
  +-AttributeDefinition[name=attr,type=Int NULL]                                                                                                                                  
  +-AttributeDefinition[name=attr,type=Float]                                                                                                                                     
==

Comment (zuyu) [inline parser/tests/Create.test]: Just write the correct query, and run the test. It should fail, but the output file should have the printed statement, then copy the statement to the test file, and it should pass.

Comment (adalbertgerald) [inline parser/tests/Create.test]: Thanks @zuyu! Can you please provide more details about the output file? Like where will it be generated and what will it be named as?

I only see the following when the test fails.

[gerald@francisco] (91)$ make test
Running tests...
Test project /scratch/gerald/quickstep/build/parser
      Start  1: quickstep_parser_tests_ParserTest_aggregate
 1/11 Test  #1: quickstep_parser_tests_ParserTest_aggregate ...   Passed    0.01 sec
      Start  2: quickstep_parser_tests_ParserTest_alter
 2/11 Test  #2: quickstep_parser_tests_ParserTest_alter .......   Passed    0.01 sec
      Start  3: quickstep_parser_tests_ParserTest_copy
 3/11 Test  #3: quickstep_parser_tests_ParserTest_copy ........   Passed    0.01 sec
      Start  4: quickstep_parser_tests_ParserTest_create
 4/11 Test  #4: quickstep_parser_tests_ParserTest_create ......***Failed    0.01 sec
      Start  5: quickstep_parser_tests_ParserTest_delete
 5/11 Test  #5: quickstep_parser_tests_ParserTest_delete ......   Passed    0.01 sec
      Start  6: quickstep_parser_tests_ParserTest_drop
 6/11 Test  #6: quickstep_parser_tests_ParserTest_drop ........   Passed    0.01 sec
      Start  7: quickstep_parser_tests_ParserTest_index
 7/11 Test  #7: quickstep_parser_tests_ParserTest_index .......   Passed    0.01 sec
      Start  8: quickstep_parser_tests_ParserTest_insert
 8/11 Test  #8: quickstep_parser_tests_ParserTest_insert ......   Passed    0.01 sec
      Start  9: quickstep_parser_tests_ParserTest_select
 9/11 Test  #9: quickstep_parser_tests_ParserTest_select ......   Passed    0.01 sec
      Start 10: quickstep_parser_tests_ParserTest_tpch
10/11 Test #10: quickstep_parser_tests_ParserTest_tpch ........   Passed    0.01 sec
      Start 11: quickstep_parser_tests_ParserTest_update
11/11 Test #11: quickstep_parser_tests_ParserTest_update ......   Passed    0.01 sec

91% tests passed, 1 tests failed out of 11

Total Test time (real) =   0.08 sec

The following tests FAILED:
          4 - quickstep_parser_tests_ParserTest_create (Failed)
Errors while running CTest
make: *** [test] Error 8

Comment (zuyu) [inline parser/tests/Create.test]: https://github.com/pivotalsoftware/quickstep/blob/master/parser/tests/CMakeLists.txt#L44

Comment (adalbertgerald) [inline parser/tests/Create.test]: Thanks @zuyu! I have added this. Also, note that this PR only adds parser support for hash partitioning.

File: parser/tests/Create.test [hunk 7]
```diff
@@ -139,6 +139,74 @@ CREATE TABLE test (attr INT) CHECK(attr>0)
...
+CreateTableStatement[relation_name=test]
++-attribute_list=
+| +-AttributeDefinition[name=attr,type=Int]
++-partition_clause=
+  +-PartitionClause[partition_type=0]
+    +-Number of Partitions=NumericLiteral[numeric_string=4,float_like=false]
+    +-attribute_name_list=
+      +-String[value=attr]
+==
+
```
Comment (zuyu) [inline parser/tests/Create.test]: We need to also test the range-based partitions.

Comment (adalbertgerald) [inline parser/tests/Create.test]: This PR adds support only for hash-based partition scheme and hence I didn't add the RANGE partition test case here. That will be part of a separate PR. Although the lexer support has been added for the RANGE keyword, parser support hasn't been added yet.

Also, can you please contact the PR owner before cancelling any builds in the future? Thanks!

Comment (zuyu) [inline parser/tests/Create.test]: OK. I've resumed the build.

Then the PR title should be `parser support for Hash Partition`.

Comment (adalbertgerald) [inline parser/tests/Create.test]: Thanks @zuyu! I have updated the PR's title.

File: parser/tests/Create.test (hunk 6 above)
Comment (jianqiao) [inline parser/tests/Create.test]: @adalbertgerald @zuyu One convenient way to print a specific test's output is, under `/build` directory, use command:
```
ctest -VV -R <regular_expression_for_test_name>
```
`-R` stands for regular expression: all tests with names that partially match `<regular_expression_for_test_name>` will be tested. The `-VV` flag is to print the full test output.

For example, in the above case:
```
ctest -VV -R quickstep_parser_tests_ParserTest_create
```

Also for example, to test all `ParseTest`'s (create, insert, etc.):
```
ctest -VV -R ParserTest
```

**Comment by zuyu** [comment]
LGTM.

File: parser/tests/Create.test (hunk 6 above)
Comment (adalbertgerald) [inline parser/tests/Create.test]: Thanks @jianqiao for pointing this out. This is exactly what I was looking for. Thanks again! :)

**Comment by adalbertgerald** [comment]
Thanks for your review comments @zuyu!

**Pull request merged**